feat(gatsby): enable granular chunks (#22253)
gatsbyjs/gatsby: docs/docs/production-app.md

      // e.g. webpack-runtime-e402cdceeae5fad2aa61.js
      name: `webpack-runtime`
    },
    splitChunks: {
      chunks: `all`,
      cacheGroups: {
        // disable Webpack's default cacheGroup
        default: false,
        // disable Webpack's default vendor cacheGroup
        vendors: false,

>         // Create a framework bundle that contains react libraries

Reviewer: ```suggestion
        // Create a framework bundle that contains React libraries
```

Reviewer: Sorry missed this last one!

>         // They hardly change so we bundle them to getter to improve caching

Reviewer: ```suggestion
        // They hardly change so we bundle them together to improve caching
```

Author: LOL 🤣  to getter XD

Reviewer: Clearly you write too much code :P

        framework: {},
        // Big modules that are over 160kb are moved to their own file to
        // optimize browser parsing & execution
        lib: {},
        // All libraries that are used on all pages are moved into a common chunk
        commons: {},
        // When a module is used more than once we create a shared bundle to save user's bandwidth
        shared: {},

>         // All css is bundled into one stylesheet

Reviewer: ```suggestion
        // All CSS is bundled into one stylesheet
```

        styles: {}
      },
      // Keep maximum initial requests to 25
      maxInitialRequests: 25,

>       // A chunk should be at least 20kb to be considered into splitChunks

Reviewer: ```suggestion
      // A chunk should be at least 20kb before using splitChunks
```

      minSize: 20000
    },
    minimizers: [
      // Minify javascript using Terser (https://terser.org/)
      plugins.minifyJs(),

>       // Minify css by using cssnano (https://cssnano.co/)

Reviewer: ```suggestion
      // Minify CSS by using cssnano (https://cssnano.co/)
```

There's a lot going on here. And this is just a sample of the output that doesn't include the loaders, rules, etc. We won't go over everything here, but most of it is geared towards proper code splitting of your application.

> The splitChunks section is the most complex part of our config as it configures how we generate the most optimized bundles for your website. We call this Granular Chunks as we try to make the generated javascript files as granular as we can by deduplicating all modules. You can read more about [SplitChunks](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks) and [chunks](https://webpack.js.org/concepts/under-the-hood/#chunks) on the [official webpack website](https://webpack.js.org/).

Reviewer: ```suggestion
The splitChunks section is the most complex part of the Gatsby webpack config as it configures how Gatsby generates the most optimized bundles for your website. This is referred to as Granular Chunks as Gatsby tries to make the generated JavaScript files as granular as possible by deduplicating all modules. You can read more about [SplitChunks](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks) and [chunks](https://webpack.js.org/concepts/under-the-hood/#chunks) on the [official webpack website](https://webpack.js.org/).
```

Author: Thanks for cleaning up 🙏❤️.

This contains the small [webpack-runtime](https://webpack.js.org/concepts/manifest/#runtime) as a separate bundle (configured in `optimization` section). In practice, the app and webpack-runtime are always needed together.

##### framework-[contenthash].js

> The framework bundle contains the react framework. We've noticed that React hardly gets upgraded to a newer version. Creating a separate bundle improves users' browser cache hit rate as this bundle is likely not going to be updated often.

Reviewer: ```suggestion
The framework bundle contains the React framework. Based on user behavior, React hardly gets upgraded to a newer version. Creating a separate bundle improves users' browser cache hit rate as this bundle is likely not going to be updated often.
```

##### commons-[contenthash].js

> Libraries used on every gatsby page are bundled into the commons javascript file. By bundling these all together, we make sure your user only needs to download this bundle once.

Reviewer: ```suggestion
Libraries used on every Gatsby page are bundled into the commons javascript file. By bundling these together, you can make sure your users only need to download this bundle once.
```

This contains the small [webpack-runtime](https://webpack.js.org/concepts/manifest/#runtime) as a separate bundle (configured in `optimization` section). In practice, the app and webpack-runtime are always needed together.

##### framework-[contenthash].js

> The framework bundle contains the React framework. Based on user behavior, React hardly gets upgraded to a newer version. Creating a separate bundle improves users' browser cache hit rate as this bundle is likely not going to be updated often.

Reviewer 2: So this would be `react` and `react-dom`?

Author: Yes and everything react needs, scheduler,...
